### ISSUE TITLE
【Critical issue】pulsar-broker block when process start

### ISSUE THREAD

**Issue by xiaotongwang1** [apache/pulsar#13362]
pulsar version : 2.8.1

cat broker.stack |grep -A 10  pulsar-web 

"pulsar-web-47-3" #235 prio=5 os_prio=0 tid=0x00007f95e2227800 nid=0x2d33b0 waiting on condition [0x00007f91693e2000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000005d7d613a0> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1707)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1742)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1947)
	at org.apache.pulsar.broker.web.PulsarWebResource.validateTopicOwnership(PulsarWebResource.java:591)
	at org.apache.pulsar.broker.admin.impl.PersistentTopicsBase.internalGetInternalStats(PersistentTopicsBase.java:1111)
--
"pulsar-web-47-2" #234 prio=5 os_prio=0 tid=0x00007f95e2226800 nid=0x2d33af waiting on condition [0x00007f91694e3000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000058b89e408> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1707)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1742)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1947)
	at org.apache.pulsar.broker.web.PulsarWebResource.validateTopicOwnership(PulsarWebResource.java:591)
	at org.apache.pulsar.broker.admin.impl.PersistentTopicsBase.internalGetInternalStats(PersistentTopicsBase.java:1111)
--
"pulsar-web-47-1" #233 prio=5 os_prio=0 tid=0x00007f95e21e5000 nid=0x2d33ae waiting on condition [0x00007f91695e4000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000586183dc0> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1707)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1742)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1947)
	at org.apache.pulsar.broker.web.PulsarWebResource.validateTopicOwnership(PulsarWebResource.java:591)
	at org.apache.pulsar.broker.admin.impl.PersistentTopicsBase.internalGetStats(PersistentTopicsBase.java:1100)

60/64 pulsar-web block on validateTopicOwnership

**Comment by xiaotongwang1** [comment]
"pulsar-io-4-35" #403 prio=5 os_prio=0 tid=0x00007fb428039000 nid=0x2e9c04 waiting on condition [0x00007fb334c65000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000005be7f4680> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1707)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1742)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at io.streamnative.pulsar.handlers.kop.KafkaTopicManager.removePersistentTopicAndReferenceProducerByRemoteAddress(KafkaTopicManager.java:410)
	at io.streamnative.pulsar.handlers.kop.KafkaTopicManager.lambda$close$9(KafkaTopicManager.java:337)
	at io.streamnative.pulsar.handlers.kop.KafkaTopicManager$$Lambda$1405/86927891.accept(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap$KeySetView.forEach(ConcurrentHashMap.java:4649)
	at io.streamnative.pulsar.handlers.kop.KafkaTopicManager.close(KafkaTopicManager.java:331)
	at io.streamnative.pulsar.handlers.kop.KafkaRequestHandler.close(KafkaRequestHandler.java:446)
	at io.streamnative.pulsar.handlers.kop.KafkaRequestHandler.channelInactive(KafkaRequestHandler.java:439)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:389)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:354)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)
	at io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:277)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:831)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

**Comment by xiaotongwang1** [comment]
From 39e546958be2bef0fa4a2f0061f2d6cc5d38af0d Mon Sep 17 00:00:00 2001
From: wWX976718 <wangxiang134@huawei.com>
Date: Wed, 13 Oct 2021 16:00:05 +0800
Subject: [PATCH] [Huawei]Fix kop producer frequent register.

Offering: BigDataPlatformComputeNGPubSub

[TicketNo:]DTS2021080621983
---
 .../pulsar/handlers/kop/KafkaRequestHandler.java   |  8 +-
 .../kop/KafkaTopicConsumerManagerCache.java        | 17 ++++
 .../pulsar/handlers/kop/KafkaTopicManager.java     | 91 +++++++++++++++++-----
 3 files changed, 97 insertions(+), 19 deletions(-)

diff --git a/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java b/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
index 4066370..002cd88 100644
--- a/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
+++ b/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java
@@ -844,7 +844,13 @@ public class KafkaRequestHandler extends KafkaCommandDecoder {
         }
         topicManager.registerProducerInPersistentTopic(partitionName, persistentTopic, principal);
         // collect metrics
-        final Producer producer = KafkaTopicManager.getReferenceProducer(partitionName);
+        Producer producer = KafkaTopicManager.getReferenceProducer(partitionName, remoteAddress);
+        if (producer == null) {
+            log.warn("Topic {} get producer is null for remoteAddress [{}], re-register producer",
+                    partitionName, remoteAddress);
+            topicManager.registerProducerInPersistentTopic(partitionName, persistentTopic, principal);
+            producer = KafkaTopicManager.getReferenceProducer(partitionName, remoteAddress);
+        }
         producer.updateRates(numMessages, byteBuf.readableBytes());
         producer.getTopic().incrementPublishCount(numMessages, byteBuf.readableBytes());
         updateProducerStats(topicPartition, numMessages, byteBuf.readableBytes());
diff --git a/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicConsumerManagerCache.java b/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicConsumerManagerCache.java
index 3e76f62..88d8de2 100644
--- a/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicConsumerManagerCache.java
+++ b/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicConsumerManagerCache.java
@@ -96,6 +96,23 @@ public class KafkaTopicConsumerManagerCache {
         });
     }
 
+    public void closeByRemoteAddress(final SocketAddress remoteAddress) {
+        cache.forEach((fullTopicName, internalMap) -> {
+            Optional.ofNullable(internalMap.get(remoteAddress)).ifPresent(future -> {
+                try {
+                    Optional.ofNullable(future.get(100, TimeUnit.MILLISECONDS))
+                            .ifPresent(KafkaTopicConsumerManager::close);
+                    internalMap.remove(remoteAddress);
+                    if (internalMap.isEmpty()) {
+                        cache.remove(fullTopicName);
+                    }
+                } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                    log.warn("[{}][{}] Failed to get TCM future when trying to close it", fullTopicName, remoteAddress);
+                }
+            });
+        });
+    }
+
     @VisibleForTesting
     public int getCount() {
         final AtomicInteger count = new AtomicInteger(0);
diff --git a/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicManager.java b/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicManager.java
index 7c448fa..c7714ff 100644
--- a/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicManager.java
+++ b/open_source/pulsar-kafka/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaTopicManager.java
@@ -15,6 +15,7 @@ package io.streamnative.pulsar.handlers.kop;
 
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
+import java.util.Map;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
@@ -22,6 +23,7 @@ import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicBoolean;
 import lombok.Getter;
 import lombok.NonNull;
@@ -54,7 +56,7 @@ public class KafkaTopicManager {
         topics = new ConcurrentHashMap<>();
     // cache for references in PersistentTopic: <topicName, producer>
     @Getter
-    private static final ConcurrentHashMap<String, Producer>
+    private static final ConcurrentHashMap<String, Map<SocketAddress, Producer>>
         references = new ConcurrentHashMap<>();
 
     private final InternalServerCnx internalServerCnx;
@@ -290,14 +292,15 @@ public class KafkaTopicManager {
             }
             return;
         }
-        if (references.containsKey(topicName)) {
+        if (references.containsKey(topicName) && references.get(topicName).containsKey(remoteAddress)) {
             return;
         }
         synchronized (this) {
-            if (references.containsKey(topicName)) {
+            if (references.containsKey(topicName) && references.get(topicName).containsKey(remoteAddress)) {
                 return;
             }
-            references.put(topicName, registerInPersistentTopic(persistentTopic, principal));
+            references.computeIfAbsent(topicName, ignored -> new ConcurrentHashMap<>())
+                .computeIfAbsent(remoteAddress, ignored -> registerInPersistentTopic(persistentTopic, principal));
         }
     }
 
@@ -310,16 +313,15 @@ public class KafkaTopicManager {
             }
             return;
         }
-
         try {
-            closeKafkaTopicConsumerManagers();
+            closeKafkaTopicConsumerManagersByRemoteAddress(remoteAddress);
 
             topics.keySet().forEach(topicName -> {
                 if (log.isDebugEnabled()) {
                     log.debug("[{}] remove producer {} for topic {} at close()",
                             requestHandler.ctx.channel(), references.get(topicName), topicName);
                 }
-                removePersistentTopicAndReferenceProducer(topicName);
+                removePersistentTopicAndReferenceProducerByRemoteAddress(topicName, remoteAddress);
             });
         } catch (Exception e) {
             log.error("[{}] Failed to close KafkaTopicManager. exception:",
@@ -327,14 +329,14 @@ public class KafkaTopicManager {
         }
     }
 
-    public static Producer getReferenceProducer(String topicName) {
-        return references.get(topicName);
+    public static Producer getReferenceProducer(String topicName, final SocketAddress remoteAddress) {
+        Map<SocketAddress, Producer> producerMap = references.getOrDefault(topicName, new ConcurrentHashMap<>());
+        return producerMap.get(remoteAddress);
     }
 
-    private static void removePersistentTopicAndReferenceProducer(final String topicName) {
+    public static void removePersistentTopicAndReferenceProducer(final String topicName) {
         // 1. Remove PersistentTopic and Producer from caches, these calls are thread safe
         final CompletableFuture<Optional<PersistentTopic>> topicFuture = topics.remove(topicName);
-        final Producer producer = references.remove(topicName);
 
         if (topicFuture == null) {
             removeTopicManagerCache(topicName);
@@ -346,13 +348,55 @@ public class KafkaTopicManager {
             // It's safe to wait until the future is completed because it's completed when
             // `BrokerService#getTopicIfExists` completed and it won't block too long.
             final Optional<PersistentTopic> persistentTopic = topicFuture.get();
-            if (producer != null && persistentTopic.isPresent()) {
-                try {
-                    persistentTopic.get().removeProducer(producer);
-                } catch (IllegalArgumentException ignored) {
-                    log.error("[{}] The producer's topic ({}) doesn't match the current PersistentTopic",
-                            topicName, (producer.getTopic() == null) ? "null" : producer.getTopic().getName());
-                }
+            persistentTopic.ifPresent(topic ->
+                Optional.ofNullable(references.remove(topicName)).ifPresent(map ->
+                    map.forEach((remoteAddress, producer) -> {
+                        if (log.isDebugEnabled()) {
+                            log.debug("[{}] Remove and close producer '{}'", topicName, producer);
+                        }
+                        try {
+                            topic.removeProducer(producer);
+                        } catch (IllegalArgumentException ignored) {
+                            log.error("[{}] The producer's topic ({}) doesn't match the current PersistentTopic",
+                                    topicName, (producer.getTopic() == null) ? "null" : producer.getTopic().getName());
+                        }
+                    })
+                )
+            );
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Failed to get topic '{}' in removeTopicAndReferenceProducer", topicName, e);
+        }
+    }
+
+    private static void removePersistentTopicAndReferenceProducerByRemoteAddress(final String topicName,
+                                                                                 final SocketAddress remoteAddress) {
+        // 1. Remove PersistentTopic and Producer from caches, these calls are thread safe
+        final CompletableFuture<Optional<PersistentTopic>> topicFuture = topics.get(topicName);
+        if (topicFuture == null) {
+            removeTopicManagerCache(topicName);
+            return;
+        }
+
+        // 2. Remove Producer from PersistentTopic's internal cache
+        try {
+            // It's safe to wait until the future is completed because it's completed when
+            // `BrokerService#getTopicIfExists` completed and it won't block too long.
+            final Optional<PersistentTopic> persistentTopic = topicFuture.get();
+            Map<SocketAddress, Producer> producerMap = references.get(topicName);
+            if(producerMap != null && persistentTopic.isPresent()){
+                Optional.ofNullable(producerMap.get(remoteAddress)).ifPresent(producer -> {
+                    try {
+                        persistentTopic.get().removeProducer(producer);
+                    } catch (IllegalArgumentException ignored) {
+                        log.error("[{}] The producer's topic ({}) doesn't match the current PersistentTopic",
+                                topicName, (producer.getTopic() == null) ? "null" : producer.getTopic().getName());
+                    }
+                    producerMap.remove(remoteAddress);
+                    if (producerMap.isEmpty()) {
+                        references.remove(topicName);
+                        topics.remove(topicName);
+                    }
+                });
             }
         } catch (InterruptedException | ExecutionException e) {
             log.error("Failed to get topic '{}' in removeTopicAndReferenceProducer", topicName, e);
@@ -364,6 +408,7 @@ public class KafkaTopicManager {
             removeTopicManagerCache(topicName);
 
             TCM_CACHE.removeAndClose(topicName);
+
             removePersistentTopicAndReferenceProducer(topicName);
         } catch (Exception e) {
             log.error("Failed to close reference for individual topic {}. exception:", topicName, e);
@@ -379,4 +424,14 @@ public class KafkaTopicManager {
         }
         TCM_CACHE.close();
     }
+
+    private static void closeKafkaTopicConsumerManagersByRemoteAddress(final SocketAddress remoteAddress) {
+        synchronized (KafkaTopicManager.class) {
+            if (cursorExpireTask != null) {
+                cursorExpireTask.cancel(true);
+                cursorExpireTask = null;
+            }
+        }
+        TCM_CACHE.closeByRemoteAddress(remoteAddress);
+    }
 }
-- 
2.6.4.windows.1

**Comment by tisonkun** [comment]
Closing as stale. If it's still relevant to maintained versions, feel free to open a new issue
